File: 3_BasicDataStructures/Python/src/tree_traversal.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.rite = None

    def add_left(self, data):
        self.left = Node(data)
        return self.left

    def add_rite(self, data):
        self.rite = Node(data)
        return self.rite

    def __radd__(self, data):
        return data + self.data

from collections import deque

def bfs(root, func):
    if root is None or func is None:
        return
    de = deque([root])
    while de:
        node = de.popleft()
        func(node)
        if node.left:
            de.append(node.left)
        if node.rite:
            de.append(node.rite)


def dfs_preorder(node, func):
    if node is None or func is None:
        return
    if func:
        func(node)
    dfs_preorder(node.left, func)
    dfs_preorder(node.rite, func)


def dfs_preorder_iter(root, func):
    if root is None or func is None:
        return
    stack = [root]
    while stack:
        node = stack.pop()
        func(node)
        if node.rite:
            stack.append(node.rite)
        if node.left:
            stack.append(node.left)


def dfs_inorder(node, func=None):
    if node is None or func is None:
        return
    dfs_inorder(node.left, func)
    func(node)
    dfs_inorder(node.rite, func)


def dfs_inorder_iter(root, func):
    stack = []
    curr = root
    while True:
        if curr:
            stack.append(curr)
            curr = curr.left
        else:
            if stack:
                curr = stack.pop()
                func(curr)
                curr = curr.rite
            else:
                break

def dfs_postorder(node, func):
    if node is None or func is None:
        return
    dfs_postorder(node.left, func)
    dfs_postorder(node.rite, func)
    func(node)


def dfs_postorder_iter(root, func):
    if root is None or func is None:
        return
    stack1 = [root]
    stack2 = []
    while stack1:
        node = stack1.pop()
        stack2.append(node)
        if node.left:
            stack1.append(node.left)
        if node.rite:
            stack1.append(node.rite)
    while stack2:
        node = stack2.pop()
        func(node)

if __name__ == "__main__":
    root = Node(1)
    n2 = root.add_left(2)
    n3 = root.add_rite(3)
    n2.add_left(4)
    n2.add_rite(5)

    fprn = lambda node: print(str(node.data))

    class Sum:
        def __init__(self):
            self.s = 0
        def __call__(self, node):
            self.s += node.data
        def __str__(self):
            return str(self.s)

    # bfs(root, fprn)

    dfs_postorder_iter(root, func=fprn)
    print("---------------")

    summ = Sum()
    dfs_inorder_iter(root, func=summ)
    print("Sum: "+str(summ))

File: 3_BasicDataStructures/Java/BasicDataStructures/src/TreeTraversal.java

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Stack;
import java.util.function.Consumer;

public class TreeTraversal {
    public static void main(String [] args){
        Node root = new Node(1);
        Node n2 = root.leftNode(2);
        Node n3 = root.riteNode(3);
        Node n4 = n2.leftNode(4);
        Node n5 = n2.riteNode(5);

        System.out.println("Bfs:");
        bfs(root, ((n) -> System.out.print(n.data() + " ")));

        System.out.println("\n\nDfs preorder recursive:");
        dfs_preorder(root, ((n) -> System.out.print(n.data() + " ")));
        System.out.println("\n\nDfs preorder iterative:");
        dfs_preorder_iter(root, ((n) -> System.out.print(n.data() + " ")));

        System.out.println("\n\nDfs postorder recursive:");
        dfs_postorder(root, ((n) -> System.out.print(n.data() + " ")));
        System.out.println("\n\nDfs postorder iterative:");
        dfs_postorder_iter(root, ((n) -> System.out.print(n.data() + " ")));

        System.out.println("\n\nDfs inorder recursive:");
        dfs_inorder(root, ((n) -> System.out.print(n.data() + " ")));
        System.out.println("\n\nDfs inorder iterative:");
        dfs_inorder_iter(root, ((n) -> System.out.print(n.data() + " ")));

        System.out.println("\n\nFinished");
    }

    public static void bfs(Node root, Consumer<Node> func){
        if (root == null)
            return;
        Deque<Node> q = new ArrayDeque<>();
        q.add(root);
        while (!q.isEmpty()){
            Node n = q.poll();
            //System.out.println(" " + n.data());
            func.accept(n);
            if(n.left() != null)
                q.addLast(n.left());
            if(n.rite() != null)
                q.addLast(n.rite());
        }
    }

    public static void dfs_preorder(Node root, Consumer<Node> func){
        if (root == null) return;
        func.accept(root);
        dfs_preorder(root.left(), func);
        dfs_preorder(root.rite(), func);
    }

    public static void dfs_preorder_iter(Node root, Consumer<Node> func) {
        if (root == null || func == null) return;
        Stack<Node> stack = new Stack<Node>();
        stack.push(root);
        while (!stack.isEmpty()) {
            Node node = stack.pop();
            func.accept(node);
            if (node.rite() != null)
                stack.push(node.rite());
            if (node.left() != null)
                stack.push(node.left());
        }
    }

    public static void dfs_postorder(Node root, Consumer<Node> func){
        if (root == null) return;
        dfs_postorder(root.left(), func);
        dfs_postorder(root.rite(), func);
        func.accept(root);
    }

    public static void dfs_postorder_iter(Node root, Consumer<Node> func) {
        if (root == null || func == null) return;
        Stack<Node> stack1 = new Stack<Node>();
        Stack<Node> stack2 = new Stack<Node>();
        stack1.push(root);
        while (!stack1.isEmpty()) {
            Node node = stack1.pop();
            stack2.push(node);
            if (node.left() != null)
                stack1.push(node.left());
            if (node.rite() != null)
                stack1.push(node.rite());
        }
        while (!stack2.isEmpty()) {
            Node node = stack2.pop();
            func.accept(node);
        }
    }


    public static void dfs_inorder(Node root, Consumer<Node> func){
        if (root == null) return;
        dfs_inorder(root.left(), func);
        func.accept(root);
        dfs_inorder(root.rite(), func);
    }

    public static void dfs_inorder_iter(Node root, Consumer<Node> func) {
        if (root == null || func == null) return;
        Stack<Node> stack = new Stack<Node>();
        Node curr = root;
        while(true){
            if (curr != null){
                stack.push(curr);
                curr = curr.left();
            } else {
                if (!stack.isEmpty()){
                    curr = stack.pop();
                    func.accept(curr);
                    curr = curr.rite();
                } else {
                    break;
                }
            }

        }
    }
}
